C-language API permits weaker backward compatibility requirements.

jhwoodyatt opened the pull request:
Change to the C-language API extracted from #1003 for introducing the CAML_API_VERSION mechanism to improve control of the backward compatibility requirement.

- Add CAML_API_VERSION parameter in the "caml/compatibility.h" header.
- Default value is 100.
- Explicitly define CAML_NAME_SPACE if CAML_API_VERSION >= 400.
- If CAML_API_VERSION < 403, then also #define caml_stat_heap_xxx_size.
- If CAML_API_VERSION < 400, then also #define Modify, Begin_roots, End_roots.
- Always include "caml/compatibility.h" from "caml/memory.h" header.
- Never include "caml/compatibility.h" from other headers.

jhwoodyatt committed:
C-language API permits weaker backward compatibility requirements.
- Add CAML_API_VERSION parameter in the "caml/compatibility.h" header.
- Default value is 100.
- Explicitly define CAML_NAME_SPACE if CAML_API_VERSION >= 400.
- If CAML_API_VERSION < 403, then also #define caml_stat_heap_xxx_size.
- If CAML_API_VERSION < 400, then also #define Modify, Begin_roots, End_roots.
- Always include "caml/compatibility.h" from "caml/memory.h" header.
- Never include "caml/compatibility.h" from other headers.

dra27 commented:
Sorry, changing the base did not do what I'd hoped - please could you rebase this commit on to trunk?

damiendoligez commented on byterun/caml/compatibility.h:
```diff
@@ -18,11 +18,24 @@
+
+#ifndef CAML_NAME_SPACE
+
+#if CAML_API_VERSION < 403
```
This is fishy: the test is redundant because lines 26--30 ensure that `CAML_API_VERSION < 400`.

jhwoodyatt replied:
p1. I'm rebasing onto trunk now.
p2. I'm moving the `!defined(CAML_NAME_SPACE)` predicate back where it was, because I think it's not meant to apply in cases like this.

damiendoligez reviewed (changes requested):
There is a problem in compatibility.h. The rest looks OK.

You also need to rebase to trunk. (and also change the base? I'm not sure)

xavierleroy commented on byterun/caml/memory.h:
```diff
@@ -465,6 +464,8 @@ CAMLextern struct caml__roots_block *caml_local_roots;  /* defined in roots.c */
 
+#if CAML_API_VERSION < 400
+
 /*
```
I'm afraid that `Begin_roots` and `End_roots` are used in code generated by the venerable CamlIDL tool (https://github.com/xavierleroy/camlidl/).  The tool hasn't been updated in a long time but still has users.  Forcing them to set `CAML_API_VERSION` to 312 or something is not nice.

jhwoodyatt replied:
Okay. Now that this is on `trunk` and not `4.07` I'll get around to this after my vacation.

jhwoodyatt replied:
So, first question about this: if the venerable *CamlIDL* doesn't set `CAML_API_VERSION` then it defaults to 100, and `Begin_roots`/`End_roots` definitions are available accordingly. Am I mistaken?

jhwoodyatt commented:
I have rebased this on trunk and addressed the "fishy" thing that @damiendoligez noticed. I'm ready to make whatever change @xavierleroy thinks is necessary to keep CamlIDL from being damaged, but I'm not sure I know what to do there.

xavierleroy commented:
My problem is that I'm not sure what CAML_API_VERSION=100 means:
1. all runtime APIs since version 1.00 can be used, or 
2. only runtime APIs that were in version 1.00 can be used.

CamlIDL is a case where we need both old APIs (for code generated by CamlIDL) and newer APIs (for code hand-written by users to do things that CamlIDL cannot handle).

jhwoodyatt commented:
> My problem is that I'm not sure what CAML_API_VERSION=100 means...

I had this problem too, and I agree it's confusing. Here's a proposal, which I hope perhaps @stedolan or somebody else more central to the multicore effort will review, that would expand the set of configuration variables used for API backward compatibility selection.

1. Change all the instances of `CAML_API_VERSION` in this change to `CAML_API_COMPATIBILITY`.
2. Introduce another configuration variable `CAML_API_VERSION` which matches the version in the top level **VERSION** file. (I might be able to write some logic that automatically generates the definition during the compiler build.)

```
#define CAML_API_COMPATIBILITY 100
#define CAML_API_VERSION 408
```

This would allow consumers of the runtime API A) to assert their minimum API compatibility using definitions of the `CAML_API_COMPATIBILITY` variable before inclusion of *<caml/compatibility.h>*, and B) to adapt their internal logic to version of the API in the current compiler switch.

jhwoodyatt commented:
Gentle acknowledgement that I'm still holding the ball on this problem, and I'm not sure what to do with it.

damiendoligez commented:
If I understand correctly, your `CAML_API_COMPATIBILITY` variable is defined by the user and is the minimum API version needed by the user's code, while the `CAML_API_VERSION` is the actual version number and defined by the runtime's headers.

jhwoodyatt commented:
> If I understand correctly, your `CAML_API_COMPATIBILITY` variable is defined by the user and is the minimum API version needed by the user's code, while the `CAML_API_VERSION` is the actual version number and defined by the runtime's headers.

That's what I think I was doing, yes. Again, I'm not sure this pull request is really the appropriate way to address the underlying problem, or even if the underlying problem is important enough to address. My interest in it developed when I tried to make some of my personal code with a few C-language foreign functions in it compile under the Multicore OCaml branch, which seems like it will need something in the runtime interface that allows extension implementations to compile according to the version of the runtime available.

jhwoodyatt commented:
When I get a few spare cycles next, I'll see about resolving those merge conflicts.

XVilka commented:
@johnwhitington will be this part of the 4.10 by any chance?

johnwhitington commented:
@XVilka I assume that was autocomplete for @jhwoodyatt :-)

XVilka commented:
Yes, sorry about that.

gadmm commented:
This is superseded by #9167 in case it is accepted. Please see the discussion there.

jhwoodyatt commented:
I'm tracking the progress of #9167 and will withdraw this if that achieves consensus.

gadmm commented:
For the record, I proposed a clarification of CAML_API_VERSION renamed as CAML_NAME_SPACE at #9167 which can be reused. (The bug #9167 was meant to fix was fixed differently but I still think that the approach there is preferable if some mechanism of the sort is deemed useful.)

jhwoodyatt commented:
Abandoning this request as the basic concern was more recently considered in #9167 and keeping this open when that has been closed seems unnecessary.
